From 5a552367893f7b048ec7857fcf08df417d5d0302 Mon Sep 17 00:00:00 2001
From: aahmed878 <Alsameema_Ahmedansar@comcast.com>
Date: Tue, 14 Mar 2023 05:50:04 +0000
Subject: [PATCH 3/3] qtdemux: Identify total samples in sbgp node and handle
 piff parsing to use sample properties kid if available

DELIA-59316 : [AAMP][Charter] Playback failure with keyrotation on HEVC streams

Reason for change: Identify total samples in sbgp node and handle piff parsing to use sample properties kid if available.
Test Procedure: Refer Ticket
Risks: Medium
Priority: P1

Signed-off-by: Alsameema.A <Alsameema_AhmedAnsar@Comcast.Com>
---
 gst/isomp4/qtdemux.c | 12 +++++++-----
 gst/isomp4/qtdemux.h |  2 +-
 2 files changed, 8 insertions(+), 6 deletions(-)

diff --git a/gst/isomp4/qtdemux.c b/gst/isomp4/qtdemux.c
index 6386790..fb119fc 100644
--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -249,7 +249,6 @@ struct _QtDemuxSampleToGroup
 {
   guint32 grouping_type;
   guint32 grouping_type_parameter;      /* version 1 only */
-
   QtDemuxSampleToGroupEntry *sbgp_entries;
   guint32 sbgp_entries_length;
 
@@ -392,6 +391,8 @@ static gchar *qtdemux_uuid_bytes_to_string (gconstpointer uuid_bytes);
 
 static GstStructure *qtdemux_get_cenc_sample_properties (GstQTDemux * qtdemux,
     QtDemuxStream * stream, guint sample_index);
+static void qtdemux_prepare_cenc_sample_properties (GstQTDemux * qtdemux,
+    QtDemuxStream * stream);
 static void gst_qtdemux_append_protection_system_id (GstQTDemux * qtdemux,
     const gchar * id);
 static void qtdemux_gst_structure_free (GstStructure * gststructure);
@@ -2599,6 +2600,7 @@ gst_qtdemux_sample_to_group_free (QtDemuxStream * stream)
         (GDestroyNotify) gst_qtdemux_sample_to_group_free_internal);
 
   stream->sample_to_group = NULL;
+  stream->n_samples_sbgp = 0;
 }
 
 
@@ -2950,7 +2952,6 @@ qtdemux_parse_piff (GstQTDemux * qtdemux, const guint8 * buffer, gint length,
   ss_info->crypto_info =
       g_ptr_array_new_full (sample_count,
       (GDestroyNotify) qtdemux_gst_structure_free);
-
   for (i = 0; i < sample_count; ++i) {
     GstStructure *properties;
     guint8 *data;
@@ -4091,7 +4092,7 @@ qtdemux_parse_senc (GstQTDemux * qtdemux, QtDemuxStream * stream, GstByteReader
   ss_info->crypto_info =
       g_ptr_array_new_full (sample_count,
       (GDestroyNotify) qtdemux_gst_structure_free);
-
+  qtdemux_prepare_cenc_sample_properties (qtdemux, stream);
   for (i = 0; i < sample_count; ++i) {
     GstStructure *properties;
     guint16 n_subsamples = 0;
@@ -4301,6 +4302,7 @@ qtdemux_parse_sbgp (GstQTDemux * qtdemux, GstByteReader * br,
 
     entry->sample_count = sample_count;
     entry->group_description_idx = group_description_idx;
+    stream->n_samples_sbgp += entry->sample_count;
   }
 
   stream->sample_to_group =
@@ -4378,7 +4380,7 @@ qtdemux_parse_sgpd (GstQTDemux * qtdemux, GstByteReader * br,
   g_return_val_if_fail (stream != NULL, FALSE);
   g_return_val_if_fail (br != NULL, FALSE);
 
-  if (!stream->n_samples_moof)
+  if (!stream->n_samples_sbgp)
     return FALSE;
 
   if (!gst_byte_reader_get_uint32_be (br, &version))
@@ -4495,6 +4497,7 @@ qtdemux_parse_moof (GstQTDemux * qtdemux, const guint8 * buffer, guint length,
     sbgp_node =
         qtdemux_tree_get_child_by_type_full (traf_node, FOURCC_sbgp,
         &sbgp_data);
+    stream->n_samples_sbgp = 0;
     while (sbgp_node) {
       qtdemux_parse_sbgp (qtdemux, &sbgp_data, stream);
       /* iterate all siblings */
@@ -4659,7 +4662,6 @@ qtdemux_parse_moof (GstQTDemux * qtdemux, const guint8 * buffer, guint length,
 
       qtdemux_parse_uuid (qtdemux, uuid_buffer, box_length);
     }
-
     /* if no new base_offset provided for next traf,
      * base is end of current traf */
     base_offset = running_offset;
diff --git a/gst/isomp4/qtdemux.h b/gst/isomp4/qtdemux.h
index 7fdeb1e..16793b1 100644
--- a/gst/isomp4/qtdemux.h
+++ b/gst/isomp4/qtdemux.h
@@ -492,7 +492,7 @@ struct _QtDemuxStream
   gboolean disabled;
   /* SampleToGroup information structure */
   GList *sample_to_group;
-
+  guint32 n_samples_sbgp;
   /* stereoscopic video streams */
   GstVideoMultiviewMode multiview_mode;
   GstVideoMultiviewFlags multiview_flags;
-- 
2.24.3

